python/GafferOSLUI/OSLObjectUI.py

```diff
@@ -41,6 +42,111 @@
 
 import GafferOSL
 
+import imath
+import functools
+
+_primitiveVariableNamesOptions = {
+	"P" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Point ),
+	"Pref" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Point ),
+	"velocity" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Vector ),
+	"uv" : IECore.V2fData( imath.V2f(), IECore.GeometricData.Interpretation.UV ),
+	"width" : IECore.FloatData(),
+	 "Cs" : IECore.Color3fData(),
+	 "customInt" : IECore.IntData(),
+	 "customFloat" : IECore.FloatData(),
+	 "customVector" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Vector ),
+	 "customNormal" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Normal ),
+	 "customPoint" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Point ),
+	 "customUV" : IECore.V2fData( imath.V2f(), IECore.GeometricData.Interpretation.UV ),
+	 "customColor" : IECore.Color3fData(),
+	 "customMatrix" : IECore.M44fData(),
+	 "customString" : IECore.StringData(),
+	 "multiInput" : None,
+}
+
+##########################################################################
+# _PrimitiveVariablesFooter
+##########################################################################
+
+class _PrimitiveVariablesFooter( GafferUI.PlugValueWidget ) :
+
+	def __init__( self, plug ) :
+
+		row = GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal )
+
+		GafferUI.PlugValueWidget.__init__( self, row, plug )
+
+		with row :
+
+				GafferUI.Spacer( imath.V2i( GafferUI.PlugWidget.labelWidth(), 1 ) )
+
+				menuButton = GafferUI.MenuButton(
+					image = "plus.png",
+					hasFrame = False,
+					menu = GafferUI.Menu(
+						Gaffer.WeakMethod( self.__menuDefinition ),
+						title = "Add Input"
+					),
+					toolTip = "Add Input"
+				)
+				menuButton.setEnabled( not Gaffer.MetadataAlgo.readOnly( plug ) )
+
+				GafferUI.Spacer( imath.V2i( 1 ), imath.V2i( 999999, 1 ), parenting = { "expand" : True } )
+
+	def _updateFromPlug( self ) :
+
+		self.setEnabled( self._editable() )
+
+	def __menuDefinition( self ) :
+
+		result = IECore.MenuDefinition()
+		usedNames = set()
+		for p in self.getPlug().children():
+			usedNames.add( p["name"].getValue() )
+
+		for label, constructor in sorted( _primitiveVariableNamesOptions.items() ):
+			if label.startswith( "custom" ):
+				if label in usedNames:
+					suffix = 2
+					while True:
+						newLabel = label + str( suffix )
+						if not newLabel in usedNames:
+							break
+						suffix += 1
+					label = newLabel
+				menuLabel = "/Custom/" + label[6:]
+			elif label == "multiInput":
+				menuLabel = "/Advanced/Multi Input"
```

Comment:
My thought was that all of OutFloat and OutObject would be removed from the menus in the name of simplicity and clarity.  If for some reason you need to do some weird technical thing ( we haven't actually thought of any concrete use cases for the closure input yet ), I was thinking you can build a closure yourself using OSLCode.

Is that too obscure?  I really hesitate to put much emphasis on the possibility of closure inputs, since as far as I know all currently existing use cases would be better served by the new interface.

Reply:
I should also say, in terms of the name, I can't really think of anything I like better, but I'm open to anything you come up with.  We just want something that says "that sounds weird and complicated, I shouldn't worry about it" to 95% of our users.

Reply:
> My thought was that all of OutFloat and OutObject would be removed from the menus in the name of simplicity and clarity.

Can you add a commit for that then? Their continued existence meant that I assumed you intended them to be used with multi inputs.

> I was thinking you can build a closure yourself using OSLCode.

You definitely could, but it would be even less obvious than the OSLObject approach we're replacing. I guess this is all a bit theoretical without a solid use case, but I'm not certain that wanting to bundle several primvars in one connection is a sign that one also wants to learn about the low level details of ObjectProcessing.h and closures. I think there's a reasonable argument that if it's visible, it should be obvious what to do with it.

> I really hesitate to put much emphasis on the possibility of closure inputs, since as far as I know all currently existing use cases would be better served by the new interface.

I'm open to keeping them out of the menu until a need arises, at which point we can consider a friendlier name/workflow.

> I can't really think of anything I like better, but I'm open to anything you come up with.

"Variable Set" perhaps? For a decent workflow I think it needs to be a name that we can also use for an OSLObject-like shader that we intend folks to connect in. I'm not feeling confident about the naming though, so if you think just omitting the menu item is OK, then let's just do that.

Reply:
Chatting offline, we agreed to rename the menu to "Advanced/Closure" and hide the Out* shaders. We're assuming that anyone who needs such an advanced feature will know a friendly local R&D person who can show them how to hook it up via OSLCode...